src/samplers/support/adapt.jl

```diff
@@ -1,32 +1,58 @@
 # Acknowledgement: this adaption settings is mimicing Stan's 3-phase adaptation.
 
+# Ref： https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_var_estimator.hpp
+mutable struct VarEstimator{T<:Real}
+  n :: Int
+  μ :: Vector{T}
+  M :: Vector{T}
+end
+
+reset!(ve::VarEstimator) = begin
+  ve.n = 0
+  ve.μ = zeros(ve.μ)
+  ve.M = zeros(ve.M)
+end
+
+add_sample!{T<:Real}(ve::VarEstimator{T}, s::Vector{T}) = begin
+  ve.n += 1
+  δ = s - ve.μ
+  ve.μ += δ / ve.n
+  ve.M += δ .* (s - ve.μ)
+end
+
+get_var(ve::VarEstimator) = begin
+  @assert ve.n >= 2
+  var = ve.M / (ve.n - 1)
+  var = (ve.n / (ve.n + 5.0)) * var + 1e-3 * (5.0 / (ve.n + 5.0))
+  return var
+end
+
+
+
 type WarmUpManager
-  iter_n    ::    Int
-  state     ::    Int
+  adapt_n   ::    Int
   params    ::    Dict
+  ve        ::    VarEstimator
 end
 
 getindex(wum::WarmUpManager, param) = wum.params[param]
 
 setindex!(wum::WarmUpManager, value, param) = wum.params[param] = value
 
 init_warm_up_params{T<:Hamiltonian}(vi::VarInfo, spl::Sampler{T}) = begin
-  wum = WarmUpManager(1, 1, Dict())
+  D = length(vi[spl])
+  ve = VarEstimator{Float64}(0, zeros(D), zeros(D))
+
+  wum = WarmUpManager(1, Dict(), ve)
 
   # Pre-cond
-  wum[:θ_num] = 0
-  wum[:θ_mean] = nothing
-  D = length(vi[spl])
-  wum[:stds] = ones(D)
-  wum[:vars] = ones(D)
+  
+  wum[:stds] = 1.0
 
-  # DA
-  wum[:ϵ] = nothing
-  wum[:μ] = nothing
-  wum[:ϵ_bar] = 1.0
-  wum[:H_bar] = 0.0
-  wum[:m] = 0
-  wum[:n_adapt] = spl.alg.n_adapt
+  # Dual averaging
+  wum[:ϵ] = []
+  reset_da(wum)
+  wum[:n_warmup] = spl.alg.n_adapt
```

Comment:
perhaps merge `spl.alg.n_adapt` and `spl.alg.delta` into `spl.info[:adapt_conf]`?